records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,187 @@
+remaining afterward. Ideally, we want to buy/save all for all of these in one
+go.
+
+[2] The EKS environment listens [on the `readinessProbe` health check](https://github.com/cds-snc/notification-manifests/blob/6ac406b315112d2b66d93d8fa74406c3e6e7639f/base/api-deployment.yaml#L216)
+that our setup exposes and it involves a direct database check. When the
+database slows down, so does the probe. The direct reaction from EKS is to
+detach the instances from the application's load balancers.
+
+_TODO: Add sequence or flow chart_
+
+## Options
+
+### Stream the notifications save
+
+Upon receiving a notification from the API, do not save it in the current
+*notification-api* process but defer it to a celery task. This does not
+address the root cause of our issue but offer many benefits, namely:
+
+1. This is an iterative improvement where we send the database save into
+   [Celery land](https://docs.celeryproject.org/en/stable/). Having these into
+   a Celery tasks means we can more easily control the flow of these operations
+   and control these with all of the Celery features. Specifically, we can batch
+   the notifications using [celery-batch](https://pypi.org/project/celery-batches/).
+2. The number of database connections will be lowered in the API component
+   as the saves moved into a Celery task.
+
+### Insert the notifications as a batch
+
+As we streamline the notification save with the previous option and got
+these under Celery land, we can now batch these easily. There is a
+[celery-batch](https://pypi.org/project/celery-batches/) project which
+exists that might offer us the option to easily achieve this with minimal
+setup.
+
+### Tweaking autovacuum options
+
+There are a few autovacuum configuration that we can tweak in order to delay
+the autovacuum action that triggers a performance slowdown of our database:
+
+* **autovacuum_scale_factor** (default: 0.2): we could use this to delay the vacuum, (suggestion 0.3).
+* **cost_delay**: We could use this to delay by increasing the threshold.
+
+### Manual execution of the vacuum process on the `notifications` table
+
+An option is to keep the autovacuum enabled but trigger a manual vacuum prior
+to traffic time. This would better prepare the database for upcoming load.
+
+Don't run `vacuum FULL`: this will take on most database resources and prevent
+proper operation of the database.
+
+### Disable the autovacuum process on the `notifications` table
+
+To avoid the autovacuum to process the `notifications` table during the daily
+blast of notifications, we could disable the autovacuum to prevent it from
+triggering. This need to be done carefully, as a maximum limit of 4 billion
+transactions can occur in PostgreSQL. The autovacuum cleanup reset this
+transaction number. If this threshold is reached, our database will be severely
+affected.
+
+This is a risky move and certainly not a long/mid term solution.
```

Comment:
Is this actually risky? How long could we go without any vacuuming before things fell over? Months? Could we just have some sort of manual vacuum overnight (or, alternatively, can we set autovacuum so that it doesn't run during business hours?)

Reply:
It's hard to tell for me without exploration. I am not sure if the transaction ID for postgres would shoot to the roof without constant autovacuum. We could know a bit more by disabling in staging.. but given the common wisdom provided to be careful with this, if we would assume we are safe and something happens, that will be critical.